File: kaji/packages/ts/src/capabilities/result.ts
import { validateArtifactRef, type ArtifactRef } from "@/artifacts/types";
import { MAX_DURABLE_TOOL_RESULT_BYTES } from "@/events/errors";
import { durableJsonSnapshot, type DeepReadonly, type JsonValue } from "@/events/json";

const CAPABILITY_RESULT = Symbol("kaji.capability-result");

export interface CapabilityResult<T extends JsonValue = JsonValue> {
  readonly [CAPABILITY_RESULT]: true;
  readonly value?: DeepReadonly<T>;
  readonly artifacts: readonly ArtifactRef[];
}

export function capabilityResult<T extends JsonValue = JsonValue>(
  value?: T,
  artifacts: readonly ArtifactRef[] = [],
): CapabilityResult<T> {
  const ids = new Set<string>();
  const validatedArtifacts = artifacts.map((item) => {
    const artifact = validateArtifactRef(item);
    if (ids.has(artifact.id)) throw new TypeError(`duplicate artifact id ${artifact.id}`);
    ids.add(artifact.id);
    return artifact;
  });
  const result: CapabilityResult<T> =
    value === undefined
      ? { [CAPABILITY_RESULT]: true, artifacts: Object.freeze(validatedArtifacts) }
      : {
          [CAPABILITY_RESULT]: true,
          value: durableJsonSnapshot(
            value,
            "tool_result",
            MAX_DURABLE_TOOL_RESULT_BYTES,
          ) as DeepReadonly<T>,
          artifacts: Object.freeze(validatedArtifacts),
        };
  return Object.freeze(result);
}

export function isCapabilityResult(value: unknown): value is CapabilityResult {
  return typeof value === "object" && value !== null && CAPABILITY_RESULT in value;
}

File: kaji/packages/py/src/capabilities/result.py
"""Validated output from a Capability execution."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Sequence

from kaji.artifacts import ArtifactRef, validate_artifact_ref
from kaji.events.errors import MAX_DURABLE_TOOL_RESULT_BYTES
from kaji.events.json import JsonValue, durable_json_snapshot


@dataclass(frozen=True)
class CapabilityResult:
    value: JsonValue | None = None
    artifacts: tuple[ArtifactRef, ...] = ()

    def __post_init__(self) -> None:
        object.__setattr__(
            self,
            "value",
            durable_json_snapshot(
                self.value,
                subject="tool_result",
                max_bytes=MAX_DURABLE_TOOL_RESULT_BYTES,
            ),
        )
        artifacts = tuple(validate_artifact_ref(item) for item in self.artifacts)
        ids = [item.id for item in artifacts]
        if len(ids) != len(set(ids)):
            raise ValueError("artifact ids must be unique within a capability result")
        object.__setattr__(self, "artifacts", artifacts)

    def to_tool_result(self) -> JsonValue:
        """Serialize this validated result for the existing tool execution path."""

        return durable_json_snapshot(
            {
                "value": self.value,
                "artifacts": [
                    artifact.model_dump(mode="json", exclude_none=True)
                    for artifact in self.artifacts
                ],
            },
            subject="tool_result",
            max_bytes=MAX_DURABLE_TOOL_RESULT_BYTES,
        )


def capability_result(
    value: JsonValue | None = None,
    artifacts: Sequence[ArtifactRef | dict[str, Any]] = (),
) -> CapabilityResult:
    return CapabilityResult(
        value=value,
        artifacts=tuple(validate_artifact_ref(item) for item in artifacts),
    )
